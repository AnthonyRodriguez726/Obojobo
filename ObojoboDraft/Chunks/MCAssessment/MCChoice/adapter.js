let Adapter = {
	construct(model, attrs) {
		if (__guard__(attrs != null ? attrs.content : undefined, x => x.score) != null) {
			model.modelState.score = attrs.content.score
			model.modelState._score = attrs.content.score
		} else {
			model.modelState.score = ''
		}
	},

	clone(model, clone) {
		clone.modelState.score = model.modelState.score
	},

	toJSON(model, json) {
		json.content.score = model.modelState.score
	}
}

export default Adapter
let __guard__ = (value, transform) => {
	return typeof value !== 'undefined' && value !== null ? transform(value) : undefined
}
