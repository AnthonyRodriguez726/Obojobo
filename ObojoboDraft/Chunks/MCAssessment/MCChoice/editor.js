import React from 'react'
import { Block } from 'slate'
import { CHILD_REQUIRED, CHILD_TYPE_INVALID } from 'slate-schema-violations'

const MCCHOICE_NODE = 'ObojoboDraft.Chunks.MCAssessment.MCChoice'
const MCANSWER_NODE = 'ObojoboDraft.Chunks.MCAssessment.MCAnswer'
const MCFEEDBACK_NODE = 'ObojoboDraft.Chunks.MCAssessment.MCFeedback'

import MCAnswer from '../MCAnswer/editor'
import MCFeedback from '../MCFeedback/editor'
import DefaultNode from '../../../../src/scripts/oboeditor/default-node'

class Feedback extends React.Component {
	delete() {
		const editor = this.props.editor
		const change = editor.value.change()
		change.removeNodeByKey(this.props.node.key)

		editor.onChange(change)
	}
	render(){
		return (
			<div className={'obojobo-draft--chunks--feedback'} {...this.props.attributes}>
				<button className={'delete'} onClick={() => this.delete()}>X</button>
				<p className={'label'} contentEditable={false}>{'Feedback'}</p>
				{this.props.children}
			</div>
		)
	}
}

class Answer extends React.Component {
	delete() {
		const editor = this.props.editor
		const change = editor.value.change()
		change.removeNodeByKey(this.props.node.key)

		editor.onChange(change)
	}
	render(){
		return (
			<div className={'obojobo-draft--chunks--answer'} {...this.props.attributes}>
				{this.props.children}
			</div>
		)
	}
}

class Node extends React.Component {
	constructor(props) {
		super(props);
		this.state = props.node.data.get('content')
	}
	delete(event) {
		event.stopPropagation()
		const editor = this.props.editor
		const change = editor.value.change()
		change.removeNodeByKey(this.props.node.key)

		editor.onChange(change)
	}
	handleScoreChange(event){
		const editor = this.props.editor
		const change = editor.value.change()

		this.setState({ score: event.target.checked ? 100 : 0 })

		change.setNodeByKey(this.props.node.key, { data: { content: {
			score: event.target.checked ? 100 : 0
		}}})
		editor.onChange(change)
	}
	addFeedback() {
		const editor = this.props.editor
		const change = editor.value.change()

		const newFeedback = Block.create({
			type: MCFEEDBACK_NODE
		})
		change.insertNodeByKey(this.props.node.key, this.props.node.nodes.size, newFeedback)

		editor.onChange(change)
	}
	render(){
		const hasFeedback = this.props.node.nodes.size === 2
		return (
			<div
				className={'component obojobo-draft--chunks--mc-assessment--mc-choice is-not-selected is-correct is-type-could-have-chosen is-mode-practice'}
				{...this.props.attributes}>
				<button className={'delete'} onClick={event => this.delete(event)}>X</button>
				<input type="checkbox" checked={this.state.score === 100} onChange={event => this.handleScoreChange(event)}/>
				<div className={'children'}>
					<div>
						{this.props.children}
					</div>
				</div>
				{!hasFeedback ? <button onClick={() => this.addFeedback()}>{'Add Feedback'}</button> : null}
			</div>
		)
	}
}

const slateToObo = node => {
	const json = {}
	json.id = node.key
	json.type = node.type
	json.content = node.data.get('content') || {}
	json.children = []

	node.nodes.forEach(child => {
		// If the current Node is a registered OboNode, use its custom converter
		if(child.type === MCANSWER_NODE || child.type === MCFEEDBACK_NODE){
			json.children.push(MCAnswer.helpers.slateToObo(child))
		} else {
			json.children.push(DefaultNode.helpers.slateToObo(child))
		}
	})

	return json
}

const oboToSlate = node => {
	const json = {}
	json.object = 'block'
	json.key = node.id
	json.type = node.type
	json.data = { content: {} }

	return json
}

const plugins = {
	renderNode(props) {
		switch (props.node.type) {
			case MCCHOICE_NODE:
				return <Node {...props} />
		}
	},
	schema: {
		blocks: {
			'ObojoboDraft.Chunks.MCAssessment.MCChoice': {
				nodes: [
					{ types: [MCANSWER_NODE], min: 1, max: 1 },
					{ types: [MCFEEDBACK_NODE], max: 1}
				],
				normalize: (change, violation, { node, child, index }) => {
					switch (violation) {
						case CHILD_REQUIRED: {
							const block = Block.create({
								type: MCANSWER_NODE
							})
							return change.insertNodeByKey(node.key, index, block)
						}
						case CHILD_TYPE_INVALID: {
							// extra children will be deleted by slate defaults
							if (index >= 2) return
							// multiple answers will be deleted by slate defaults
							if (index === 1 && child.type !== MCFEEDBACK_NODE) return
							return change.wrapBlockByKey(
								child.key,
								{
									type: MCANSWER_NODE
								}
							)
						}
					}
				}
			}
		}
	}
}

const MCChoice = {
	components: {
		Node,
		Answer,
		Feedback
	},
	helpers: {
		slateToObo,
		oboToSlate,
	},
	plugins
}

export default MCChoice
