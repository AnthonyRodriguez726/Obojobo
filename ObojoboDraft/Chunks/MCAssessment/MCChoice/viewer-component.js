import './viewer-component.scss'

import Common from 'Common'
import Viewer from 'Viewer'
import isOrNot from '../../../../src/scripts/common/isornot'

let { OboComponent } = Common.components
let { OboModel } = Common.models

let { QuestionUtil } = Viewer.util

const getInputType = responseType => {
	switch (responseType) {
		case 'pick-all':
			return 'checkbox'
		case 'pick-one':
		case 'pick-one-multiple-correct':
		default:
			return 'radio'
	}
}

const questionIsSelected = (questionState, model) => {
	let response = QuestionUtil.getResponse(
		questionState,
		model.getParentOfType('ObojoboDraft.Chunks.Question')
	) || { ids: [] }

	return response.ids.indexOf(model.get('id')) !== -1
}

const answerIsCorrect = (model, mode, questionState, navStateContext) => {
	let score
	if (mode === 'review') {
		// no score data for this context? no idea what to do, throw an error
		if (!questionState.scores[navStateContext]) throw 'Unkown Question State'

		score = QuestionUtil.getScoreForModel(questionState, model, navStateContext)
	} else {
		score = model.modelState.score
	}

	return score === 100
}

const MCChoice = props => {
	let isCorrect

	try {
		isCorrect = answerIsCorrect(
			props.model,
			props.mode,
			props.moduleData.questionState,
			props.moduleData.navState.context
		)
	} catch (error) {
		// if there's no questionState data for this
		// or getting the score throws an error
		// just display a div
		return <div />
	}

	let isSelected = questionIsSelected(props.moduleData.questionState, props.model)

	let className =
		'obojobo-draft--chunks--mc-assessment--mc-choice' +
		isOrNot(isSelected, 'selected') +
		isOrNot(isCorrect, 'correct') +
		' is-mode-' +
		props.mode

	return (
		<OboComponent
			model={props.model}
			moduleData={props.moduleData}
			className={className}
			data-choice-label={props.label}
		>
			<input
				type={getInputType(props.responseType)}
				value={props.model.get('id')}
				checked={isSelected}
				name={props.model.parent.get('id')}
			/>
			<div className="children">
				{props.model.children.map((child, index) => {
					let type = child.get('type')
					let isAnswerItem = type === 'ObojoboDraft.Chunks.MCAssessment.MCAnswer'
					let isFeedbackItem = type === 'ObojoboDraft.Chunks.MCAssessment.MCFeedback'
					if (isAnswerItem) {
						let Component = child.getComponentClass()
						return <Component key={child.get('id')} model={child} moduleData={props.moduleData} />
					}
				})}
			</div>
		</OboComponent>
	)
}

MCChoice.defaultProps = {
	responseType: null,
	revealAll: false,
	questionSubmitted: false
}

export default MCChoice
