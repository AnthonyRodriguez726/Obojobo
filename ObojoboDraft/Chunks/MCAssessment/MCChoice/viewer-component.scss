@import '~styles/includes';

.obojobo-draft--chunks--mc-assessment--mc-choice {
	list-style-type: none;
	min-height: 3.5em;
	position: relative;
	transition: background 0.1s ease;
	font-family: $font-default;
	padding-bottom: 1em;
	cursor: pointer;
	overflow: auto;

	input {
		font-size: inherit;
		position: absolute;
		top: 1.7em;
		left: 3em;
		transform: translate(0, -50%) translate(0, 0.5px);
		margin: 0;
	}

	.children {
		margin-left: 2em;
		margin-right: 0;

		.feedback-icon {
			position: absolute;
			opacity: 0.4;
			width: 1em;
			right: 0.5em;
			top: 50%;
			transform: translate(0, -50%);
		}
	}

	label,
	.label {
		cursor: pointer;
		display: table;
		min-height: 2em;
		padding: 0.5em 0;
		position: relative;
		width: 100%;
	}

	.input {
		display: table-cell;
		padding-left: 2em;
		vertical-align: middle;
		width: 100%;

		&:focus {
			box-shadow: none;
			outline: none;

			.text {
				box-shadow: $box-shadow-focus;
			}
		}
	}

	.feedback {
		box-sizing: border-box;
		color: transparentize($color-text, 0.3);
		font-size: 0.7em;
		margin-top: -0.5em;
		padding: 0.1rem 2em;
		padding-bottom: 0.5em;
		position: relative;

		.placeholder {
			left: 0.5em;
		}
	}

	&.is-selected {
		background: transparentize($color-highlight, 0.75);
	}

	&.is-mode-review {
		input {
			display: none;
		}
		.children {
			margin-left: 0;
		}
		&:hover {
			cursor: default;
			background: inherit;
		}
		cursor: default;
	}

	&:hover {
		background: transparentize($color-highlight, 0.85);
	}

	.correct-button {
		position: relative;
		top: 1em;
		left: 1em;
		font-size: inherit;
		background: transparent;
		border: none;
	}
	.add-feedback {
		float: right;
		font-size: 0.85em;
		background-color: $color-bg2;
		border: 1px solid transparentize($color-shadow, 0.2);
		border-radius: $dimension-rounded-radius;
		margin: 0.2em;

		&:hover {
			background-color: darken($color-bg2, 20%);
		}
	}

	&.is-correct {
		.correct-button {
			color: $color-correct;

			&:hover {
				color: darken($color-correct, 20%);
			}
		}
	}

	&.is-not-correct {
		.correct-button {
			color: $color-incorrect;
			&:hover {
				color: darken($color-incorrect, 20%);
			}
		}
	}
}
