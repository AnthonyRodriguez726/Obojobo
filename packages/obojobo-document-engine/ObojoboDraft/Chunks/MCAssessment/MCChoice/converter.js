import Common from 'Common'

const slateToObo = node => ({
	id: node.key,
	type: node.type,
	children: node.nodes.map(child => Common.Store.getItemForType(child.type).slateToObo(child)),
	content: node.data.get('content') || {}
})

const oboToSlate = node => ({
	object: 'block',
	key: node.id,
	type: node.type,
	nodes: node.children.map(child => Common.Store.getItemForType(child.type).oboToSlate(child)),
	data: {
		content: node.content
	}
})

export default { slateToObo, oboToSlate }
