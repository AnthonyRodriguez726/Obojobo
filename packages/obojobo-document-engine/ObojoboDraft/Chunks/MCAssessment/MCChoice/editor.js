import React from 'react'
import { Block } from 'slate'
import { CHILD_REQUIRED, CHILD_TYPE_INVALID } from 'slate-schema-violations'

const MCCHOICE_NODE = 'ObojoboDraft.Chunks.MCAssessment.MCChoice'
const MCANSWER_NODE = 'ObojoboDraft.Chunks.MCAssessment.MCAnswer'
const MCFEEDBACK_NODE = 'ObojoboDraft.Chunks.MCAssessment.MCFeedback'

import MCAnswer from '../MCAnswer/editor'
import DefaultNode from '../../../../src/scripts/oboeditor/components/default-node'
import isOrNot from '../../../../src/scripts/common/isornot'

import Common from 'Common'

class Node extends React.Component {
	constructor(props) {
		super(props)
	}

	delete(event) {
		event.stopPropagation()
		const editor = this.props.editor
		const change = editor.value.change()
		change.removeNodeByKey(this.props.node.key)

		editor.onChange(change)
	}

	handleScoreChange(event) {
		event.stopPropagation()
		const editor = this.props.editor
		const change = editor.value.change()
		const newScore = this.props.node.data.get('content').score === 100 ? 0 : 100

		change.setNodeByKey(this.props.node.key, {
			data: {
				content: {
					score: newScore
				}
			}
		})
		editor.onChange(change)
	}

	addFeedback() {
		const editor = this.props.editor
		const change = editor.value.change()

		const newFeedback = Block.create({
			type: MCFEEDBACK_NODE
		})
		change.insertNodeByKey(this.props.node.key, this.props.node.nodes.size, newFeedback)

		editor.onChange(change)
	}

	render() {
		const score = this.props.node.data.get('content').score
		const hasFeedback = this.props.node.nodes.size === 2
		const className =
			'component obojobo-draft--chunks--mc-assessment--mc-choice' +
			isOrNot(score === 100, 'correct')
		return (
			<div className={className}>
				<button className={'delete-node'} onClick={event => this.delete(event)}>
					X
				</button>
				<button className={'correct-button'} onClick={event => this.handleScoreChange(event)}>
					{score === 100 ? '✔' : '✖'}
				</button>
				<div className={'children'}>
					<div>{this.props.children}</div>
				</div>
				{!hasFeedback ? (
					<button className={'add-feedback'} onClick={() => this.addFeedback()}>
						{'Add Feedback'}
					</button>
				) : null}
			</div>
		)
	}
}

const slateToObo = node => {
	const json = {}
	json.id = node.key
	json.type = node.type
	json.content = node.data.get('content') || {}
	json.children = []

	node.nodes.forEach(child => {
		// If the current Node is a registered OboNode, use its custom converter
		switch (child.type) {
			case MCANSWER_NODE:
				json.children.push(MCAnswer.helpers.slateToObo(child))
				break
			case MCFEEDBACK_NODE:
				json.children.push(MCAnswer.helpers.slateToObo(child))
				break
			default:
				json.children.push(DefaultNode.helpers.slateToObo(child))
		}
	})

	return json
}

const oboToSlate = node => {
	const json = {}
	json.object = 'block'
	json.key = node.id
	json.type = node.type
	json.data = { content: node.content }
	json.nodes = []

	node.children.forEach(child => {
		// If the current Node is a registered OboNode, use its custom converter
		switch (child.type) {
			case MCANSWER_NODE:
				json.nodes.push(MCAnswer.helpers.oboToSlate(child))
				break
			case MCFEEDBACK_NODE:
				json.nodes.push(MCAnswer.helpers.oboToSlate(child))
				break
			default:
				json.nodes.push(DefaultNode.helpers.oboToSlate(child))
		}
	})

	return json
}

const plugins = {
	renderNode(props) {
		switch (props.node.type) {
			case MCCHOICE_NODE:
				return <Node {...props} {...props.attributes} />
		}
	},
	schema: {
		blocks: {
			'ObojoboDraft.Chunks.MCAssessment.MCChoice': {
				nodes: [
					{
						match: [{ type: MCANSWER_NODE }],
						min: 1,
						max: 1
					},
					{
						match: [{ type: MCFEEDBACK_NODE }],
						max: 1
					}
				],
				normalize: (change, error) => {
					const { node, child, index } = error
					switch (error.code) {
						case CHILD_REQUIRED: {
							const block = Block.create({
								type: MCANSWER_NODE
							})
							return change.insertNodeByKey(node.key, index, block)
						}
						case CHILD_TYPE_INVALID: {
							// extra children will be deleted by slate defaults
							if (index >= 2) return
							// multiple answers and feedbacks will be deleted by slate defaults
							if (index === 1 && child.type !== MCFEEDBACK_NODE) return
							return change.wrapBlockByKey(child.key, {
								type: MCANSWER_NODE
							})
						}
					}
				}
			}
		}
	}
}

const MCChoice = {
	components: {
		Node
	},
	helpers: {
		slateToObo,
		oboToSlate
	},
	plugins
}

Common.Store.registerEditorModel('ObojoboDraft.Chunks.MCAssessment.MCChoice', {
	name: 'Multiple Choice Choice',
	isInsertable: false,
	slateToObo,
	oboToSlate,
	plugins
})

export default MCChoice
