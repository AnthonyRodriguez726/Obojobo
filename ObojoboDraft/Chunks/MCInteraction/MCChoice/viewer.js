import Common from 'Common'
import adapter from './adapter'
import ViewerComponent from './viewer-component'

let SelectionHandler = Common.chunk.textChunk.TextGroupSelectionHandler

Common.Store.registerModel('ObojoboDraft.Chunks.MCInteraction.MCChoice', {
	type: 'chunk',
	adapter: adapter,
	componentClass: ViewerComponent,
	selectionHandler: new SelectionHandler()
})
