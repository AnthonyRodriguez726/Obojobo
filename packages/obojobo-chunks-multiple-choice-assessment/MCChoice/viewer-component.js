import './viewer-component.scss'

import React from 'react'

import Common from 'obojobo-document-engine/src/scripts/common/index'
import Viewer from 'obojobo-document-engine/src/scripts/viewer/index'
import isOrNot from 'obojobo-document-engine/src/scripts/common/isornot'

const { OboComponent } = Common.components
const { QuestionUtil } = Viewer.util

const QUESTION_TYPE = 'ObojoboDraft.Chunks.Question'
const CHOSEN_CORRECTLY = 'chosen-correctly'
const SHOULD_NOT_HAVE_CHOSEN = 'should-not-have-chosen'
const COULD_HAVE_CHOSEN = 'could-have-chosen'
const SHOULD_HAVE_CHOSEN = 'should-have-chosen'
const UNCHOSEN_CORRECTLY = 'unchosen-correctly'

const getInputType = responseType => {
	switch (responseType) {
		case 'pick-all':
			return 'checkbox'
		case 'pick-one':
		case 'pick-one-multiple-correct':
		default:
			return 'radio'
	}
}

const questionIsSelected = (questionState, model, navStateContext) => {
	const response = QuestionUtil.getResponse(
		questionState,
		model.getParentOfType(QUESTION_TYPE),
		navStateContext
	) || { ids: [] }

	return response.ids.indexOf(model.get('id')) !== -1
}

const getQuestionModel = model => model.getParentOfType(QUESTION_TYPE)

const answerIsCorrect = (model, mode, questionState, navStateContext) => {
	let score
	if (mode === 'review') {
		// no score data for this context? no idea what to do, throw an error
		if (!questionState.scores[navStateContext]) throw 'Unkown Question State'

		score = QuestionUtil.getScoreForModel(questionState, getQuestionModel(model), navStateContext)
	} else {
		score = model.modelState.score
	}
	return score === 100
}

const renderAnsFlag = type => {
	let flagEl

	switch (type) {
		case UNCHOSEN_CORRECTLY:
			return <div />
		case CHOSEN_CORRECTLY:
			flagEl = <p>Your Answer (Correct)</p>
			break
		case SHOULD_NOT_HAVE_CHOSEN:
			flagEl = <p>Your Answer (Incorrect)</p>
			break
		case COULD_HAVE_CHOSEN:
			flagEl = <p>Another Correct Answer</p>
			break
		case SHOULD_HAVE_CHOSEN:
			flagEl = <p> Correct Answer </p>
			break
	}

	return (
		<div className={`obojobo-draft--chunks--mc-assessment--mc-choice--answer-flag is-type-${type}`}>
			{flagEl}
		</div>
	)
}

const getAnsType = (model, isCorrect, isSelected) => {
	// The user selected a correct answer (not necessarily this one)
	// On multi-select questions, this is only true if a user selected all and only correct answers
	// Renamed for clarity w/ isACorrectChoice
	const userIsCorrect = isCorrect

	const isACorrectChoice = model.get('content').score === 100

	if (isSelected) {
		return isACorrectChoice ? CHOSEN_CORRECTLY : SHOULD_NOT_HAVE_CHOSEN
	}

	if (isACorrectChoice) {
		return userIsCorrect ? COULD_HAVE_CHOSEN : SHOULD_HAVE_CHOSEN
	}

	return UNCHOSEN_CORRECTLY
}

const MCChoice = props => {
	let isCorrect

	try {
		isCorrect = answerIsCorrect(
			props.model,
			props.mode,
			props.moduleData.questionState,
			props.moduleData.navState.context
		)
	} catch (error) {
		// if there's no questionState data for this
		// or getting the score throws an error
		// just display a div
		return <div />
	}

	const isSelected = questionIsSelected(
		props.moduleData.questionState,
		props.model,
		props.moduleData.navState.context
	)

	const ansType = getAnsType(props.model, isCorrect, isSelected)

	let flag
	if (props.mode === 'review') {
		flag = renderAnsFlag(ansType)
	}

	const className =
		'obojobo-draft--chunks--mc-assessment--mc-choice' +
		isOrNot(isSelected, 'selected') +
		isOrNot(isCorrect, 'correct') +
		` is-type-${ansType}` +
		` is-mode-${props.mode}`

	return (
		<OboComponent
			model={props.model}
			moduleData={props.moduleData}
			className={className}
			data-choice-label={props.label}
		>
			<input
				type={getInputType(props.responseType)}
				value={props.model.get('id')}
				checked={isSelected}
				name={props.model.parent.get('id')}
			/>
			<div className="children">
				{props.model.children.map(child => {
					const type = child.get('type')
					const isAnswerItem = type === 'ObojoboDraft.Chunks.MCAssessment.MCAnswer'
					const id = child.get('id')

					if (isAnswerItem) {
						const Component = child.getComponentClass()
						return (
							<div key={id}>
								{flag}
								<Component key={id} model={child} moduleData={props.moduleData} />
							</div>
						)
					}

					return null
				})}
			</div>
		</OboComponent>
	)
}

MCChoice.defaultProps = {
	responseType: null,
	revealAll: false,
	questionSubmitted: false
}

export default MCChoice
