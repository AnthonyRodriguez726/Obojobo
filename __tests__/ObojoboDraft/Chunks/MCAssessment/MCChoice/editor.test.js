import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import { CHILD_REQUIRED, CHILD_TYPE_INVALID } from 'slate-schema-violations'

jest.mock('../../../../../ObojoboDraft/Chunks/MCAssessment/MCAnswer/editor')
jest.mock('../../../../../ObojoboDraft/Chunks/MCAssessment/MCFeedback/editor')

import MCChoice from '../../../../../ObojoboDraft/Chunks/MCAssessment/MCChoice/editor'
const MCCHOICE_NODE = 'ObojoboDraft.Chunks.MCAssessment.MCChoice'
const MCFEEDBACK_NODE = 'ObojoboDraft.Chunks.MCAssessment.MCFeedback'
const MCANSWER_NODE = 'ObojoboDraft.Chunks.MCAssessment.MCAnswer'

describe('MCChoice editor', () => {
	test('Node component', () => {
		const Node = MCChoice.components.Node
		const component = renderer.create(
			<Node
				attributes={{ dummy: 'dummyData' }}
				node={{
					data: {
						get: () => {
							return {}
						}
					},
					nodes: []
				}}
			/>
		)
		const tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})

	test('Node component correct choice', () => {
		const Node = MCChoice.components.Node
		const component = renderer.create(
			<Node
				attributes={{ dummy: 'dummyData' }}
				node={{
					data: {
						get: () => {
							return { score: 100 }
						}
					},
					nodes: { size: 2 }
				}}
			/>
		)
		const tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})

	test('Node component deletes itself', () => {
		const Node = MCChoice.components.Node

		const change = {
			removeNodeByKey: jest.fn()
		}

		const component = mount(
			<Node
				node={{
					key: 'mockKey',
					nodes: [],
					data: {
						get: () => {
							return {}
						}
					}
				}}
				editor={{
					value: { change: () => change },
					onChange: jest.fn()
				}}
			/>
		)
		const tree = component.html()

		component
			.find('button')
			.at(0)
			.simulate('click')

		expect(change.removeNodeByKey).toHaveBeenCalled()
		expect(tree).toMatchSnapshot()
	})

	test('Node component changes score', () => {
		const Node = MCChoice.components.Node

		const change = {
			setNodeByKey: jest.fn()
		}

		const component = mount(
			<Node
				node={{
					key: 'mockKey',
					nodes: [],
					data: {
						get: () => {
							return {
								score: 0
							}
						}
					}
				}}
				editor={{
					value: { change: () => change },
					onChange: jest.fn()
				}}
			/>
		)

		component
			.find('button')
			.at(1)
			.simulate('click')

		expect(change.setNodeByKey).toHaveBeenCalledWith('mockKey', {
			data: { content: { score: 100 } }
		})

		const component2 = mount(
			<Node
				node={{
					key: 'mockKey',
					nodes: [],
					data: {
						get: () => {
							return {
								score: 100
							}
						}
					}
				}}
				editor={{
					value: { change: () => change },
					onChange: jest.fn()
				}}
			/>
		)

		component2
			.find('button')
			.at(1)
			.simulate('click')

		expect(change.setNodeByKey).toHaveBeenCalledWith('mockKey', {
			data: { content: { score: 0 } }
		})
	})

	test('Node component adds feedback', () => {
		const Node = MCChoice.components.Node

		const change = {
			insertNodeByKey: jest.fn()
		}

		const component = mount(
			<Node
				node={{
					key: 'mockKey',
					nodes: [],
					data: {
						get: () => {
							return {}
						}
					}
				}}
				editor={{
					value: { change: () => change },
					onChange: jest.fn()
				}}
			/>
		)

		component
			.find('button')
			.at(2)
			.simulate('click')

		expect(change.insertNodeByKey).toHaveBeenCalled()
	})

	test('slateToObo converts a Slate node to an OboNode with content', () => {
		const slateNode = {
			key: 'mockKey',
			type: 'mockType',
			data: {
				get: () => null
			},
			nodes: [
				{
					type: MCANSWER_NODE
				},
				{
					type: MCFEEDBACK_NODE
				},
				{
					type: 'notADefinedNode'
				}
			]
		}
		const oboNode = MCChoice.helpers.slateToObo(slateNode)

		expect(oboNode).toMatchSnapshot()
	})

	test('oboToSlate converts an OboNode to a Slate node', () => {
		const oboNode = {
			id: 'mockKey',
			type: 'mockType',
			children: [
				{
					type: MCANSWER_NODE
				},
				{
					type: MCFEEDBACK_NODE
				},
				{
					type: 'notADefinedNode'
				}
			]
		}
		const slateNode = MCChoice.helpers.oboToSlate(oboNode)

		expect(slateNode).toMatchSnapshot()
	})

	test('plugins.renderNode renders a choice', () => {
		const props = {
			node: {
				type: MCCHOICE_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(MCChoice.plugins.renderNode(props)).toMatchSnapshot()
	})

	test('plugins.schema.normalize fixes invalid children', () => {
		const change = {
			wrapBlockByKey: jest.fn()
		}

		MCChoice.plugins.schema.blocks[MCCHOICE_NODE].normalize(change, CHILD_TYPE_INVALID, {
			node: {},
			child: { key: 'mockKey', type: MCFEEDBACK_NODE },
			index: 1
		})

		expect(change.wrapBlockByKey).toHaveBeenCalled()
	})

	test('plugins.schema.normalize fixes too many children', () => {
		const change = {
			wrapBlockByKey: jest.fn()
		}

		MCChoice.plugins.schema.blocks[MCCHOICE_NODE].normalize(change, CHILD_TYPE_INVALID, {
			node: {},
			child: { key: 'mockKey' },
			index: 3
		})

		expect(change.wrapBlockByKey).not.toHaveBeenCalled()
	})

	test('plugins.schema.normalize fixes extra children', () => {
		const change = {
			wrapBlockByKey: jest.fn()
		}

		MCChoice.plugins.schema.blocks[MCCHOICE_NODE].normalize(change, CHILD_TYPE_INVALID, {
			node: {},
			child: { key: 'mockKey', type: 'wrongType' },
			index: 1
		})

		expect(change.wrapBlockByKey).not.toHaveBeenCalled()
	})

	test('plugins.schema.normalize adds missing children', () => {
		const change = {
			insertNodeByKey: jest.fn()
		}

		MCChoice.plugins.schema.blocks[MCCHOICE_NODE].normalize(change, CHILD_REQUIRED, {
			node: {},
			child: null,
			index: 0
		})

		expect(change.insertNodeByKey).toHaveBeenCalled()
	})
})
