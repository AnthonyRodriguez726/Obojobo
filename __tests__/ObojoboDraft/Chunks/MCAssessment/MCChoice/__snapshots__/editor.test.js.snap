// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MCChoice editor Node component 1`] = `
<div
  className="component obojobo-draft--chunks--mc-assessment--mc-choice is-not-correct"
  dummy="dummyData"
>
  <button
    className="delete"
    onClick={[Function]}
  >
    X
  </button>
  <button
    className="correct-button"
    onClick={[Function]}
  >
    ✖
  </button>
  <div
    className="children"
  >
    <div>
      mockChildren
    </div>
  </div>
  <button
    onClick={[Function]}
  >
    Add Feedback
  </button>
</div>
`;

exports[`MCChoice editor Node component adds feedback 1`] = `"<div class=\\"component obojobo-draft--chunks--mc-assessment--mc-choice is-not-correct\\"><button class=\\"delete\\">X</button><button class=\\"correct-button\\">✖</button><div class=\\"children\\"><div>mockChildren</div></div><button>Add Feedback</button></div>"`;

exports[`MCChoice editor Node component changes score 1`] = `"<div class=\\"component obojobo-draft--chunks--mc-assessment--mc-choice is-not-correct\\"><button class=\\"delete\\">X</button><button class=\\"correct-button\\">✖</button><div class=\\"children\\"><div>mockChildren</div></div><button>Add Feedback</button></div>"`;

exports[`MCChoice editor Node component correct choice 1`] = `
<div
  className="component obojobo-draft--chunks--mc-assessment--mc-choice is-correct"
  dummy="dummyData"
>
  <button
    className="delete"
    onClick={[Function]}
  >
    X
  </button>
  <button
    className="correct-button"
    onClick={[Function]}
  >
    ✔
  </button>
  <div
    className="children"
  >
    <div>
      mockChildren
    </div>
  </div>
</div>
`;

exports[`MCChoice editor Node component deletes itself 1`] = `"<div class=\\"component obojobo-draft--chunks--mc-assessment--mc-choice is-not-correct\\"><button class=\\"delete\\">X</button><button class=\\"correct-button\\">✖</button><div class=\\"children\\"><div>mockChildren</div></div><button>Add Feedback</button></div>"`;

exports[`MCChoice editor oboToSlate converts an OboNode to a Slate node 1`] = `
Object {
  "data": Object {
    "content": undefined,
  },
  "key": "mockKey",
  "nodes": Array [
    undefined,
    undefined,
    Object {
      "data": Object {
        "content": undefined,
      },
      "key": undefined,
      "object": "block",
      "type": "notADefinedNode",
    },
  ],
  "object": "block",
  "type": "mockType",
}
`;

exports[`MCChoice editor plugins.renderNode renders a choice 1`] = `
<Node
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Chunks.MCAssessment.MCChoice",
    }
  }
/>
`;

exports[`MCChoice editor slateToObo converts a Slate node to an OboNode with content 1`] = `
Object {
  "children": Array [
    undefined,
    undefined,
    Object {
      "children": Array [],
      "id": undefined,
      "type": "notADefinedNode",
    },
  ],
  "content": Object {},
  "id": "mockKey",
  "type": "mockType",
}
`;
